### PR TITLE
DGtal on MS Windows

### PR DESCRIPTION
**Pull request by dcoeurjo**
This PR fixes plenty of issues to compile DGtal on VisualStudio + MSWindows.
So far:
- [x] VisualStudio 14 (2015) 
  - [x] DGtal core, no deps: everything compiles (examples+unittests), only 5 unit tests fails or segfault at runtime
  - [x] DGtal core  + Eigen

   ~~- [ ] DGtal core + GMP~~
  ~~- [ ] DGtal core + QGLVIEWER (and QT)~~

~~- [ ] VisualStudio 12 (2013) ~~
  ~~- [ ] DGtal core, no deps: bug in Display3D template parameters..~~
 ~~- [ ] DGtal core + GMP + Eigen~~
 ~~- [ ] DGtal core + QGLVIEWER (and QT)~~
 
- [x] Test if DGtalTools compiles against MS-DGtal

For VS14, the only thing you have to do is to enable c++11 (```WITH_C11```) and to disable shared lib build (```BUILD_SHARED_LIBS=false```).

I'll create issues (or will send private emails) to DGtal developers to help me to fix some unit test issues and tricky warnings raised by windows ```cl``` compiler (which sometimes make sense).

:octocat: :sunglasses:

### CODE REVIEW

**Comment by dcoeurjo** [comment]
(PR descriptions edited)

**Comment by dcoeurjo** [comment]
PR description edited... I give up with VS2012.. Only focus on VS14

**Comment by kerautret** [comment]
for DGTalTools it is near to be fine ! I am just going to check for 64bits but done for 32 bits ;)

File: appveyor.yml [hunk 1]
```diff
@@ -1,3 +1,59 @@
...
-build:
-  verbosity: minimal
+version: 0.9beta.{build}
+
+environment:
+  matrix:
+  - VS_GEN: Visual Studio 14 2015
+    CONFIG: RelWithDebInfo
+    B_NAME: Win32
+  - VS_GEN: Visual Studio 14 2015 Win64
```
Comment (kerautret) [inline appveyor.yml]: the second configuration fails on my appveoyr account I suppose also for you ?

File: examples/tutorial-examples/imageSetDT.cpp [hunk 2]
```diff
@@ -101,8 +102,8 @@ int main()
  
 
   //! [ImageSetDT-DTvis]
-  DTL2::Value maxDT = (*std::max_element(dt.constRange().begin(),
```
Comment (kerautret) [inline examples/tutorial-examples/imageSetDT.cpp]: the std::max fails on VS ?

File: src/DGtal/geometry/curves/ArithmeticalDSS.h [hunk 3]
```diff
@@ -830,7 +830,10 @@ namespace DGtal
     public ArithmeticalDSS<TCoordinate, TInteger, 4>
   {
   public: 
-    typedef ArithmeticalDSS<TCoordinate, TInteger, 4> Super; 
+    typedef ArithmeticalDSS<TCoordinate, TInteger, 4> Super;
+	typedef typename Super::Point Point;
```
Comment (kerautret) [inline src/DGtal/geometry/curves/ArithmeticalDSS.h]: indent

File: src/DGtal/geometry/curves/ArithmeticalDSL.h [hunk 4]
```diff
@@ -778,6 +780,8 @@ namespace DGtal
      * Type of base class.
      */
     typedef ArithmeticalDSL<TCoordinate, TInteger, 8> Super;
+	typedef typename Super::Coordinate Coordinate;
```
Comment (kerautret) [inline src/DGtal/geometry/curves/ArithmeticalDSL.h]: indent

File: src/DGtal/geometry/curves/ArithmeticalDSL.h [hunk 5]
```diff
@@ -712,6 +712,8 @@ namespace DGtal
      * Type of base class.
      */
     typedef ArithmeticalDSL<TCoordinate, TInteger, 4> Super;
+	typedef typename Super::Coordinate Coordinate;
```
Comment (kerautret) [inline src/DGtal/geometry/curves/ArithmeticalDSL.h]: indent

File: src/DGtal/geometry/curves/ArithmeticalDSS.h [hunk 6]
```diff
@@ -966,7 +969,10 @@ namespace DGtal
     public ArithmeticalDSS<TCoordinate, TInteger, 8>
   {
   public: 
-    typedef ArithmeticalDSS<TCoordinate, TInteger, 8> Super; 
+    typedef ArithmeticalDSS<TCoordinate, TInteger, 8> Super;
+	typedef typename Super::Point Point;
```
Comment (kerautret) [inline src/DGtal/geometry/curves/ArithmeticalDSS.h]: indent

File: appveyor.yml (hunk 1 above)
Comment (dcoeurjo) [inline appveyor.yml]: Yep it fails too

File: src/DGtal/geometry/curves/ArithmeticalDSSComputer.h [hunk 7]
```diff
@@ -466,7 +466,8 @@ namespace DGtal
     /**
      * Base class
      */
-    typedef ArithmeticalDSSComputer<TIterator, TInteger, 4> Super; 
+    typedef ArithmeticalDSSComputer<TIterator, TInteger, 4> Super;
+	typedef typename Super::ConstIterator ConstIterator;
```
Comment (kerautret) [inline src/DGtal/geometry/curves/ArithmeticalDSSComputer.h]: indent

File: src/DGtal/geometry/curves/ArithmeticalDSSComputer.h [hunk 8]
```diff
@@ -527,6 +528,7 @@ namespace DGtal
      * Base class
      */
     typedef ArithmeticalDSSComputer<TIterator, TInteger, 4> Super; 
+	typedef typename Super::ConstIterator ConstIterator;
```
Comment (kerautret) [inline src/DGtal/geometry/curves/ArithmeticalDSSComputer.h]: idem

File: examples/tutorial-examples/imageSetDT.cpp (hunk 2 above)
Comment (dcoeurjo) [inline examples/tutorial-examples/imageSetDT.cpp]: Don't know... This is fine for me on my VM

File: src/DGtal/images/Morton.h [hunk 9]
```diff
@@ -80,12 +80,13 @@ namespace DGtal
     typedef THashKey HashKey;
     typedef TPoint Point;
     typedef typename Point::Coordinate Coordinate;
-    static const Dimension dimension ;    
 
     BOOST_CONCEPT_ASSERT(( concepts::CUnsignedNumber<THashKey> ));
     BOOST_CONCEPT_ASSERT(( concepts::CIntegralNumber<THashKey> ));
     BOOST_CONCEPT_ASSERT(( concepts::CInteger<Coordinate> ));
 
+	BOOST_STATIC_CONSTANT(Dimension, dimension = TPoint::dimension);
```
Comment (kerautret) [inline src/DGtal/images/Morton.h]: idem

File: src/DGtal/io/readers/VolReader.ih [hunk 10]
```diff
@@ -36,8 +36,7 @@
 // Interface - public :
 
 
-template <typename TImageContainer, typename TFunctor>
-const int DGtal::VolReader<TImageContainer,TFunctor>::MAX_HEADERNUMLINES = 64;
+#define MAX_HEADERNUMLINES 64
```
Comment (kerautret) [inline src/DGtal/io/readers/VolReader.ih]: the previous def cause pb in VS ? the #define can be dangerous but if it fix the compilation..

**Comment by kerautret** [comment]
all looks fine for me, nice work! (just see some typo indents).
Perhaps in the doc we can mentions that the build is always tested on VS 14 ? (like here http://liris.cnrs.fr/dgtal/doc/nightly/moduleBuildDGtal.html) and on dgtal.org ?
Else also fix the merge and then we can merge I think!

**Comment by dcoeurjo** [comment]
It has been manually tested for VS14. However, appeveyor is crashing and cannot enable it by default.

**Comment by dcoeurjo** [comment]
Furthermore I think that recent PR from Jaco makes my build crash as well ((template chec on Containers). I have to investigate

**Comment by kerautret** [comment]
It looks fine on my appveyor when testing DGTalTools from your branch merged from the current master (which contains cubical complex PR) at least in Win64 ;)

**Comment by kerautret** [comment]
but all options/deps in the build are not tested

**Comment by dcoeurjo** [comment]
Yes but  everything isn't tested in DGtaltools ;)
Great news that it's working on 64bits. 

(Note that there are still ~10 unit tests that compile but fail at runtime. )

**Comment by kerautret** [comment]
oups I was thinking a least on Win32 ;(

**Comment by dcoeurjo** [comment]
done. thanks for the comments.
Waiting for Travis now

**Comment by kerautret** [comment]
perfect travis looks fine, merging..

**Pull request merged**